### PR TITLE
Implement changes for meta-ros layer version 3 (part 1)

### PR DESCRIPTION
**Pull request by shr-project**

### CODE REVIEW

**Comment by shr-project** [comment]
I've updated this to include fixed version of all 3 open PRs from Herb and 1 from me. This branch is what we're using in meta-ros layer version 3 now.

**Comment by herb-kuta-lge** [comment]
@shr-project < Please defer https://github.com/ros-infrastructure/superflore/pull/256/commits/7820ceebb56b24e5b77281fba361db0d6db1866e until we have a complete solution for building ROS 1 components for ROS 2 distros.

**Comment by shr-project** [comment]
> @shr-project < Please defer [7820cee](https://github.com/ros-infrastructure/superflore/commit/7820ceebb56b24e5b77281fba361db0d6db1866e) until we have a complete solution for building ROS 1 packages for ROS 2 distros.

But this fixes the parsing issues and allows to start building and testing. Without this we'll be again blocked until superflore is updated and all distros regenerated to start working on this.

With the number for recipe re-generations and updates we have in queue for various milestones it would be really useful to have this already included to be able to work on ROS1+ROS2 in parallel to other initiatives.

**Comment by herb-kuta-lge** [comment]
Won't it have to be changed anyway to set the other generated variables correctly? Some of them will need to be appended, but others, eg, ROS_VERSION will need `?=`. Wouldn't it be better to make all of the changes to support ROS 1 + ROS 2 at the same time once we've decided what this will be? 

> But this fixes the parsing issues and allows to start building and testing.

You can always add a TEMP commit to make the changes you need to fix the parsing issues.

**Comment by herb-kuta-lge** [comment]
> `generators/bitbake/yocto_recipe.py:539:        conf_file_name = 'ros-distro.inc'`

This needs to be changed to `'superflore-ros-distro.inc'` and the `/generated` subdirectory added to the line 1 comment, eg. `# melodic/generated/superflore-ros-distro.inc`.

**Comment by shr-project** [comment]
> Won't it have to be changed anyway to set the other generated variables correctly? Some of them will need to be appended, but others, eg, ROS_VERSION will need `?=`. Wouldn't it be better to make all of the changes to support ROS 1 + ROS 2 at the same time once we've decided what this will be?

Without this not even bitbake -e doesn't work, because it doesn't parse due to missing native dependencies. What's wrong with solving this incrementally? The main fix would be to change the order in which ros-distro.inc files are parsed - this shouldn't need a change in superflore. At least it would be good enough to investigate what else needs to change with bitbake -e, but if you reject this one, then the very first step would need to be this superflore change or similar TEMP commit just to get started.

**Comment by herb-kuta-lge** [comment]
> What's wrong with solving this incrementally? 

Nothing, I suppose, but then place it in a separate PR so that this one contains only the changes required to implement part 1 of meta-ros layer version 3 (including the one that generates `superflore-ros-distro.inc` instead of `ros-distro.inc` mentioned above). And change the title of this PR to "Implement changes for meta-ros layer version 3 (part 1)". @allenh1 < The PR merge policy that's being used for this repo is "rebase and squash", right? (I don't see where this is shown in the UI.) So the title of the PR matters because it is what will appear as the commit summary line.

**Comment by shr-project** [comment]
Added another PR https://github.com/ros-infrastructure/superflore/pull/259 which includes all of this PR + that one commit on top.

**Comment by herb-kuta-lge** [comment]
> Added another PR #259 which includes all of this PR + that one commit on top.

**Comment by herb-kuta-lge** [comment]
Please squash "yocto_recipe.py: rename the generated ros-distro.inc to superflore-ro" into "Generate recipes in meta-ros-<distro> instead of generated-recipes-<d…"

File: superflore/generators/bitbake/yocto_recipe.py [hunk 1]
```diff
@@ -612,8 +612,12 @@ def generate_ros_distro_inc(
                     + 'tools.\n')
                 conf_file.write(
                     yoctoRecipe.generate_multiline_variable(
-                        'ROS_SUPERFLORE_GENERATED_BUILDTOOLS',
+                        'ROS_SUPERFLORE_GENERATED_BUILDTOOLS_%s' %
+                        distro.upper(),
                         yoctoRecipe.generated_native_recipes) + '\n')
+                conf_file.write('ROS_SUPERFLORE_GENERATED_BUILDTOOLS_append ='
+                                '"${ROS_SUPERFLORE_GENERATED_BUILDTOOLS_%s}"'
```
Comment (herb-kuta-lge) [inline superflore/generators/bitbake/yocto_recipe.py]: Add leading space before the $ "just in case".

**Comment by herb-kuta-lge** [comment]
> Added another PR #259 which includes all of this PR + that one commit on top.
31b491d747c0fda5f4ce9fa694c93454794b4c98

**Comment by herb-kuta-lge** [comment]
> Added another PR #259 which includes all of this PR + that one commit on top.

31b491d747c0fda5f4ce9fa694c93454794b4c98 still appears in the list shown by https://github.com/ros-infrastructure/superflore/pull/256/commits . Is this correct?

**Comment by allenh1** [comment]
@shr-project please resolve the merge conflicts, and I'll give it a look

**Comment by shr-project** [comment]
> @shr-project please resolve the merge conflicts, and I'll give it a look

The merge conflicts are caused by other 3 changes from other open PRs which I've included yesterday to resolve the merge conflicts :). Once the https://github.com/ros-infrastructure/superflore/pull/233 https://github.com/ros-infrastructure/superflore/pull/243 are merged, I'll rebase this and only last 3 commits should remain without any conflicts.

**Comment by shr-project** [comment]
> > @shr-project please resolve the merge conflicts, and I'll give it a look
> 
> The merge conflicts are caused by other 3 changes from other open PRs which I've included yesterday to resolve the merge conflicts :). Once the #233 #243 are merged, I'll rebase this and only last 3 commits should remain without any conflicts.

Thanks for merging the other 2, I've rebased this one (and the follow-up one) now.

**Pull request merged**